src/phys/dca_step/cluster_solver/shared_tools/accumulation/tp/tp_accumulator_kernels.cu

```diff
@@ -0,0 +1,434 @@
+
+      const CudaComplex<Real> Ga_2 = cond_conj(G_down[i_a + ldgd * j_a], conj_a);
+      const CudaComplex<Real> Gb_2 = cond_conj(G_up[i_b + ldgu * j_b], conj_b);
+
+      contribution = (Ga_1 * Gb_1 + Ga_2 * Gb_2) * (-factor);
+    } break;
+
+    case PARTICLE_HOLE_MAGNETIC: {
+      int w1_a(w1), w2_a(w2);
+      const bool conj_a = helper.extendWIndices(w1_a, w2_a);
+      const int i_a = b1 + nb * k1 + no * w1_a;
+      const int j_a = b4 + nb * k2 + no * w2_a;
+
+      int w1_b(helper.addWex(w2, w_ex));
+      int w2_b(helper.addWex(w1, w_ex));
+      const bool conj_b = helper.extendWIndices(w1_b, w2_b);
+      const int i_b = b2 + nb * helper.addKex(k2, k_ex) + no * w1_b;
+      const int j_b = b3 + nb * helper.addKex(k1, k_ex) + no * w2_b;
+
+      const CudaComplex<Real> Ga_1 = cond_conj(G_up[i_a + ldgu * j_a], conj_a);
+      const CudaComplex<Real> Gb_1 = cond_conj(G_up[i_b + ldgu * j_b], conj_b);
+
+      const CudaComplex<Real> Ga_2 = cond_conj(G_down[i_a + ldgd * j_a], conj_a);
+      const CudaComplex<Real> Gb_2 = cond_conj(G_down[i_b + ldgd * j_b], conj_b);
+
+      contribution = (Ga_1 * Gb_1 + Ga_2 * Gb_2) * (-factor);
+    }
+      {
+        int w1_a(w1);
+        int w2_a(helper.addWex(w1, w_ex));
+        const bool conj_a = helper.extendWIndices(w1_a, w2_a);
+        const int i_a = b1 + nb * k1 + no * w1_a;
+        const int j_a = b3 + nb * helper.addKex(k1, k_ex) + no * w2_a;
+
+        int w1_b(helper.addWex(w2, w_ex));
+        int w2_b(w2);
+        const bool conj_b = helper.extendWIndices(w1_b, w2_b);
+        const int i_b = b2 + nb * helper.addKex(k2, k_ex) + no * w1_b;
+        const int j_b = b4 + nb * k2 + no * w2_b;
+
+        const CudaComplex<Real> Ga =
+            cond_conj(G_up[i_a + ldgu * j_a] - G_down[i_a + ldgd * j_a], conj_a);
+        const CudaComplex<Real> Gb =
+            cond_conj(G_up[i_b + ldgu * j_b] - G_down[i_b + ldgd * j_b], conj_b);
+
+        contribution += (Ga * Gb) * factor;
+      }
+      break;
+
+    case PARTICLE_HOLE_CHARGE: {
+      int w1_a(w1), w2_a(w2);
+      const bool conj_a = helper.extendWIndices(w1_a, w2_a);
+      const int i_a = b1 + nb * k1 + no * w1_a;
+      const int j_a = b4 + nb * k2 + no * w2_a;
+
+      int w1_b(helper.addWex(w2, w_ex));
+      int w2_b(helper.addWex(w1, w_ex));
+      const bool conj_b = helper.extendWIndices(w1_b, w2_b);
+      const int i_b = b2 + nb * helper.addKex(k2, k_ex) + no * w1_b;
+      const int j_b = b3 + nb * helper.addKex(k1, k_ex) + no * w2_b;
+
+      const CudaComplex<Real> Ga_1 = cond_conj(G_up[i_a + ldgu * j_a], conj_a);
+      const CudaComplex<Real> Gb_1 = cond_conj(G_up[i_b + ldgu * j_b], conj_b);
+
+      const CudaComplex<Real> Ga_2 = cond_conj(G_down[i_a + ldgd * j_a], conj_a);
+      const CudaComplex<Real> Gb_2 = cond_conj(G_down[i_b + ldgd * j_b], conj_b);
+
+      contribution = (Ga_1 * Gb_1 + Ga_2 * Gb_2) * (-factor);
+    }
+      {
```

Comment:
Ok so this is essentially UpdateG4SpinDifference, but the documentation here is insufficient. This does not really follow the form of tp_accumulator.hpp too clearly for someone who doesn't dream in CUDA. Each case needs to be documented at least to the standard in the other accumulator files. Additionally what purpose are the {} serving here?  Do this instead of working on more new things and I think we will be ready to merge.  Removing the code duplication should wait until later PR after this one is merged.